lib/vrl/stdlib/src/insert.rs

```diff
@@ -0,0 +1,140 @@
+                result: Ok(r#"{ "foo": { "bar": [null, "baz"] } }"#),
+            },
+            Example {
+                title: "external target",
+                source: indoc! {r#"
+                    . = { "foo": true }
+                    insert!(., "bar", "baz")
+                "#},
+                result: Ok(r#"{ "foo": true, "bar": "baz" }"#),
+            },
+            Example {
+                title: "variable",
+                source: indoc! {r#"
+                    var = { "foo": true }
+                    insert!(var, "bar", "baz")
+                "#},
+                result: Ok(r#"{ "foo": true, "bar": "baz" }"#),
+            },
+            Example {
+                title: "invalid segment type",
+                source: r#"insert!({"foo": { "bar": [92, 42] }}, ["foo", true], "baz")"#,
+                result: Err(
+                    r#"function call error for "insert" at (0:59): expected "string" or "integer", got "boolean""#,
+                ),
+            },
+        ]
+    }
+
+    fn compile(&self, _state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let target = arguments.required("target");
+        let field = arguments.required("field");
+        let value = arguments.required("value");
+
+        Ok(Box::new(InsertFn {
+            target,
+            field,
+            value,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct InsertFn {
+    target: Box<dyn Expression>,
+    field: Box<dyn Expression>,
+    value: Box<dyn Expression>,
+}
+
+impl Expression for InsertFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let path = match self.field.resolve(ctx)? {
+            Value::Bytes(path) => LookupBuf::from(String::from_utf8_lossy(&path)),
+            Value::Array(paths) => {
+                let mut lookup = LookupBuf::root();
+
+                for path in paths {
+                    let segment = match path {
+                        Value::Bytes(path) => {
+                            SegmentBuf::Field(String::from_utf8_lossy(&path).into_owned().into())
+                        }
+                        Value::Integer(index) => SegmentBuf::Index(index as isize),
+                        value => {
+                            return Err(value::Error::Expected {
+                                got: value.kind(),
+                                expected: Kind::Bytes | Kind::Integer,
+                            }
+                            .into())
+                        }
+                    };
+
+                    lookup.push_back(segment)
+                }
+
+                lookup
+            }
+            value => {
+                return Err(value::Error::Expected {
+                    got: value.kind(),
+                    expected: Kind::Array | Kind::Bytes,
+                }
+                .into())
+            }
+        };
+
+        let mut value = self.target.resolve(ctx)?;
```

Comment:
```suggestion
        let mut target = self.target.resolve(ctx)?;
```

Minutely confusing with lots of different `value` variables flying around..